File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/moderation/ModerationDtos.kt
package dev.stackverse.backend.moderation

import com.fasterxml.jackson.annotation.JsonInclude
import dev.stackverse.backend.bookmark.BookmarkStatus
import java.time.Instant
import java.util.UUID

data class ReportRequest(val reason: String? = null, val comment: String? = null)

data class ReportResolutionRequest(val resolution: String? = null, val note: String? = null)

data class BookmarkStatusRequest(val status: BookmarkStatus? = null, val note: String? = null)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class ReportResponse(
    val id: UUID,
    val bookmarkId: UUID,
    val reporter: String,
    val reason: ReportReason,
    val comment: String?,
    val status: ReportStatus,
    val resolvedBy: String?,
    val resolvedAt: Instant?,
    val resolutionNote: String?,
    val createdAt: Instant,
) {
    companion object {
        fun of(report: Report) = ReportResponse(
            id = report.id,
            bookmarkId = report.bookmarkId,
            reporter = report.reporter,
            reason = report.reason,
            comment = report.comment,
            status = report.status,
            resolvedBy = report.resolvedBy,
            resolvedAt = report.resolvedAt,
            resolutionNote = report.resolutionNote,
            createdAt = report.createdAt,
        )
    }
}

File: backends/dotnet/src/StackverseBackend/Moderation/ModerationDtos.cs
using StackverseBackend.Bookmarks;

namespace StackverseBackend.Moderation;

// `reason` and `resolution` stay strings so an unknown value becomes a field-level
// validation error with its `validation.*` key, not a JSON parse failure.
public sealed record ReportRequest(string? Reason = null, string? Comment = null);

public sealed record ReportResolutionRequest(string? Resolution = null, string? Note = null);

public sealed record BookmarkStatusRequest(BookmarkStatus? Status = null, string? Note = null);

public sealed record ReportResponse(
    Guid Id,
    Guid BookmarkId,
    string Reporter,
    ReportReason Reason,
    string? Comment,
    ReportStatus Status,
    string? ResolvedBy,
    DateTime? ResolvedAt,
    string? ResolutionNote,
    DateTime CreatedAt)
{
    public static ReportResponse Of(Report report) => new(
        report.Id,
        report.BookmarkId,
        report.Reporter,
        report.Reason,
        report.Comment,
        report.Status,
        report.ResolvedBy,
        report.ResolvedAt,
        report.ResolutionNote,
        report.CreatedAt);
}
